### ISSUE TITLE
How to get graphs in distance mode and delta

### ISSUE THREAD

**Issue by MaxZec** [theOehrly/Fast-F1#66]
Hi there,

I was wondering if it's possibile to draw graphs using data from real F1 in distance mode rather than time mode...this would make laps comparison much more accurate, and also it would give us the chance to calculate a proper delta which is impossibile in time mode...like in the image attached here, it would looks like that one is braking\accelerating later than the other, but we can't be sure coz the graph and data are in time mode and not distance mode....and is there any chance to have some data where corners are so we can easily add T1 T2 etc rather than do it manually? Thank you in advance!
![Senza titolo-3](https://user-images.githubusercontent.com/95567617/144742352-fb0fdd68-6156-4f53-8676-22fc3b69ca97.jpg)

**Comment by theOehrly** [comment]
Yes, you can plot data over distance. I thought there was an example for that, but I can't find it. So I guess there's actually no example for it.

The recommended way is this
```
lap = laps.pick_driver('VER').pick_fastest()
tel = lap.get_car_data().add_distance()
plt.plot(tel['Distance'], tel['Speed'])
```

Parts of this article in medium.com might be interesting:
https://medium.com/towards-formula-1-analysis/how-i-analyze-formula-1-data-with-python-2021-italian-gp-dfb11db4b73


There is no data on corner numbers. It's currently not possible to add automated labels for that.

**Comment by MaxZec** [comment]
Thank you for your reply! I've done a test but I believe it's not useful for what I meant. As the data record starts in different points (it seems), the X axis (that should be the same for everybody) won't match...see the image...first data recorded for VER lap is different from HAM's one...so I can't use distance data to have a distance mode or calculate a time delta...
![Immagine](https://user-images.githubusercontent.com/95567617/144748417-be4e7d8a-71ce-4221-8c31-6120f7cdc0d4.jpg)
..

**Comment by theOehrly** [comment]
I don't quite understand what you want. Do you want the distance to start a zero and then increment in constant intervals?

**Comment by MaxZec** [comment]
Something like that yes, unless I'm missing something that can make me use that difference distances values. I'll try to make myself clear sorry.

So, if I have to compare let's say how and where HAM and VER  brake in corner 1, especially in terms of meters before the corner...to do this I need that all the data (brakes, throttle, rpm, gears etc) started from the same point and being constant through all the record data.. So let's say from Mt 5 after the start line and increment of 2 meters every record data.. So that I'm sure that if HAM brakes at 50mt and and VER a bit later means he brakes earlier...in time mode it's impossible to know this...

**Comment by theOehrly** [comment]
What you want to do is possible in theory, of course. But FastF1 does not provide any means for you to do it. The reasons for this is the low sampling rate and comparably low accuracy of the data (look at the speed traces, they are jittery).
The distance is calculated by integrating speed over time. The result is good enough for plotting but not good enough to base any further calculations on. Especially not if they require high accuracy. Determining braking points requires very high accuracy.
If you were to resample the distance values into constant intervals, the result would be highly inaccurate. You are starting with data that is not perfectly accurate, and then you introduce more inaccuracy through resampling.

Furthermore, let me give you a short calculation as an example for why the data is not accurate enough to determine braking points. 
The source data has a sample rate of 5 Hz (actually slightly lower even). That means, we have one sample every 200 milliseconds.
Assume that the car is driving at 300 km/h. That is equivalent to 83 m/s. Now, let's assume, that your data is shifted by only one sample or 1/5th of a second. 83 m/s times 0.2 seconds is 17 m. That is a huge difference!
The data is not perfectly synchronized. So you will have some inaccuracies there. The comparison that you want to do is not possible with the data we have. We'd need some well synchronized data at a considerably higher sampling rate, but the api does not provide that.

My current assumption is, that the accuracy is about +-0.05 seconds (there are outliers).

**Comment by MaxZec** [comment]
Thank you very much for clarify!!!

**Comment by OmaMansou** [comment]
Hello Max, 
I think I know what you want to do, cause I tried doing the same, but it doesn't seem like there's a solution that gives 100% accuracy.

One method I found (and that gave satisfactory results) was merging the tables on timings (for driver_1 and driver_2) based on what time is closest using pandas.merge_asof and then applying a smoothing. Here is how i did it : 

`   times_1 = telemetry_1[['Time', 'Distance']].rename(columns={'Time': 'time_1'})
    times_2 = telemetry_2[['Time', 'Distance']].rename(columns={'Time': 'time_2'})
    times = merge_asof(left=times_1,
                                     right=times_2,
                                     on='Distance',
                                     direction='nearest')
    times.loc[:, 'delta'] = times.time_2 - times.time_1
    times.loc[:, 'smooth_delta'] = times.delta.rolling(20, center=True).mean()`

The delta column is very jittery, but once the smoothing is applied (with rolling and mean), the plot looks just fine and is quite usable.

**Comment by OmaMansou** [comment]
I managed to get this plot for the qualifying session of the 2021 Bahrain GP.

![image](https://user-images.githubusercontent.com/42304770/148252016-cd84cee1-fbd5-47dc-981c-939a002389eb.png)
